backend/tournesol/views/contributor_recommendations.py

```diff
@@ -0,0 +1,119 @@
+"""
+Overrides the Polls API for recommendations specific to one user
+"""
+from django.db.models import Case, F, Prefetch, Q, Sum, When
+from django.shortcuts import get_object_or_404
+from rest_framework import serializers
+from rest_framework.permissions import IsAuthenticated
+
+from core.models import User
+from tournesol.serializers.contributor_recommendations import ContributorRecommendationsSerializer
+
+from ..models import ContributorRating, ContributorRatingCriteriaScore, Entity, Poll
+from ..views import PollRecommendationsBaseAPIView
+
+
+class ContributorRecommendationsBaseView(PollRecommendationsBaseAPIView):
+    """
+    A base class used to factorize behaviours common to all contributor
+    recommendations views.
+    """
+
+    queryset = Entity.objects.none()
+    serializer_class = ContributorRecommendationsSerializer
+
+    def annotate_with_total_score(self, queryset, request, poll: Poll, user):
+        criteria_cases = []
+        for crit in poll.criterias_list:
+            url_weight = request.query_params.get(f"weights[{crit}]")
+            if url_weight is not None:
+                try:
+                    used_weight = int(url_weight)
+                except ValueError as error_value:
+                    raise serializers.ValidationError(
+                        f"Invalid weight value for criteria '{crit}'"
+                    ) from error_value
+            else:
+                used_weight = 10
+            criteria_cases.append(
+                When(
+                    contributorvideoratings__criteria_scores__criteria=crit,
+                    then=used_weight,
+                )
+            )
+        criteria_weight = Case(*criteria_cases, default=0)
+
+        queryset = queryset.prefetch_related(
+            Prefetch(
+                "contributorvideoratings",
+                queryset=ContributorRating.objects.filter(poll=poll, user=user),
+            )
+        )
+
+        queryset = queryset.prefetch_related(
+            Prefetch(
+                "contributorvideoratings__criteria_scores",
+                queryset=ContributorRatingCriteriaScore.objects.filter(
+                    contributor_rating__poll=poll, contributor_rating__user=user
+                ),
+            )
+        )
+
+        return queryset.annotate(
+            total_score=Sum(
+                F("contributorvideoratings__criteria_scores__score") * criteria_weight,
+            ),
+            filter=Q(contributorvideoratings__poll=poll)
+            & Q(contributorvideoratings__user=user),
```

Comment:
This filter doesn't seem necessary to me, as the contributor ratings are already filtered by the prefetch on `contributorvideoratings`.

By inspecting the generated SQL, adding this filter only adds `tournesol_contributorrating.poll_id` to the SQL GROUP BY clause.

The tests still pass without this filter.

Generated SQL:
- with the filter: https://pastebin.com/raw/qA24VxX1
- without the filter: https://pastebin.com/raw/CLxHaXJF